### PR TITLE
Add option to run tests with code coverage

### PR DESCRIPTION
**Pull request by craddm**
## :white_check_mark: Checklist

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).
- [ ] You have marked this pull request as a **draft** and added `'[WIP]'` to the title if needed (if you're not yet ready to merge).

### :arrow_heading_up: Summary

~~Adds a second test function to produce code coverage as part of the test output - can now run `hatch run test:test-coverage` as well as `hatch run test:test`. The output now includes the style of coverage report below~~

`hatch run` can now be followed 

- `test:test` to run tests only
- `test:test-report` to produce a coverage report (note, `test:test` must be run before this)
- `test:test-coverage` to run both tests and produce the coverage report

The coverage report now omits the test files from the coverage percentages by default. `test-report` also now accepts additional command line arguments so the output can be customised as you wish. e.g. `hatch run test:test-report -m` will produce a report that also tells you which lines are not covered by tests.

```
Name                                                                            Stmts   Miss  Cover
---------------------------------------------------------------------------------------------------
data_safe_haven/__init__.py                                                         2      0   100%
data_safe_haven/administration/__init__.py                                          0      0   100%
data_safe_haven/administration/users/__init__.py                                    2      0   100%
data_safe_haven/administration/users/azure_ad_users.py                             42     27    36%
data_safe_haven/administration/users/guacamole_users.py                            24     13    46%
data_safe_haven/administration/users/research_user.py                              30     19    37%
data_safe_haven/administration/users/user_handler.py                              105     83    21%
data_safe_haven/commands/__init__.py                                                2      0   100%
data_safe_haven/commands/admin.py                                                  24      5    79%
data_safe_haven/commands/admin_add_users.py                                        18     11    39%
data_safe_haven/commands/admin_list_users.py                                       17     11    35%
data_safe_haven/commands/admin_register_users.py                                   30     22    27%
data_safe_haven/commands/admin_remove_users.py                                     18     12    33%
data_safe_haven/commands/admin_unregister_users.py                                 31     23    26%
data_safe_haven/commands/cli.py                                                    34     14    59%
data_safe_haven/commands/config.py                                                 29      0   100%
data_safe_haven/commands/context.py                                                64      0   100%
data_safe_haven/commands/deploy.py                                                 70     47    33%
data_safe_haven/commands/teardown.py                                               47     34    28%
data_safe_haven/config/__init__.py                                                  4      0   100%
data_safe_haven/config/config.py                                                  104      7    93%
data_safe_haven/config/pulumi.py                                                   39      0   100%
data_safe_haven/config/pulumi_project.py                                           12      2    83%
data_safe_haven/context/__init__.py                                                 3      0   100%
data_safe_haven/context/context.py                                                 60      1    98%
data_safe_haven/context/context_settings.py                                        83      6    93%
data_safe_haven/context_infrastructure/__init__.py                                  2      0   100%
data_safe_haven/context_infrastructure/infrastructure.py                           40     28    30%
data_safe_haven/exceptions/__init__.py                                             28      0   100%
data_safe_haven/external/__init__.py                                                8      0   100%
data_safe_haven/external/api/__init__.py                                            0      0   100%
data_safe_haven/external/api/azure_api.py                                         443    377    15%
data_safe_haven/external/api/azure_cli.py                                          47     24    49%
data_safe_haven/external/api/graph_api.py                                         445    392    12%
data_safe_haven/external/interface/__init__.py                                      0      0   100%
data_safe_haven/external/interface/azure_authenticator.py                          47     28    40%
data_safe_haven/external/interface/azure_container_instance.py                     56     39    30%
data_safe_haven/external/interface/azure_fileshare.py                              64     47    27%
data_safe_haven/external/interface/azure_ipv4_range.py                             34     23    32%
data_safe_haven/external/interface/azure_postgresql_database.py                   117     81    31%
data_safe_haven/functions/__init__.py                                               3      0   100%
data_safe_haven/functions/miscellaneous.py                                         16     11    31%
data_safe_haven/functions/strings.py                                               44     23    48%
data_safe_haven/infrastructure/__init__.py                                          2      0   100%
data_safe_haven/infrastructure/common/__init__.py                                   4      0   100%
data_safe_haven/infrastructure/common/enums.py                                     51      0   100%
data_safe_haven/infrastructure/common/ip_ranges.py                                 20     14    30%
data_safe_haven/infrastructure/common/transformations.py                           50     36    28%
data_safe_haven/infrastructure/components/__init__.py                               4      0   100%
data_safe_haven/infrastructure/components/composite/__init__.py                     6      0   100%
data_safe_haven/infrastructure/components/composite/automation_dsc_node.py         28     17    39%
data_safe_haven/infrastructure/components/composite/local_dns_record.py            16     10    38%
data_safe_haven/infrastructure/components/composite/microsoft_sql_database.py      24     16    33%
data_safe_haven/infrastructure/components/composite/postgresql_database.py         24     16    33%
data_safe_haven/infrastructure/components/composite/virtual_machine.py             77     55    29%
data_safe_haven/infrastructure/components/dynamic/__init__.py                       8      0   100%
data_safe_haven/infrastructure/components/dynamic/azuread_application.py           77     55    29%
data_safe_haven/infrastructure/components/dynamic/blob_container_acl.py            41     26    37%
data_safe_haven/infrastructure/components/dynamic/compiled_dsc.py                  36     16    56%
data_safe_haven/infrastructure/components/dynamic/dsh_resource_provider.py         35     20    43%
data_safe_haven/infrastructure/components/dynamic/file_share_file.py               72     48    33%
data_safe_haven/infrastructure/components/dynamic/file_upload.py                   48     30    38%
data_safe_haven/infrastructure/components/dynamic/remote_script.py                 32     17    47%
data_safe_haven/infrastructure/components/dynamic/ssl_certificate.py               91     66    27%
data_safe_haven/infrastructure/components/wrapped/__init__.py                       3      0   100%
data_safe_haven/infrastructure/components/wrapped/automation_account.py            20      7    65%
data_safe_haven/infrastructure/components/wrapped/log_analytics_workspace.py       17      6    65%
data_safe_haven/infrastructure/programs/__init__.py                                 3      0   100%
data_safe_haven/infrastructure/programs/declarative_shm.py                         22      7    68%
data_safe_haven/infrastructure/programs/declarative_sre.py                         50     32    36%
data_safe_haven/infrastructure/programs/shm/__init__.py                             0      0   100%
data_safe_haven/infrastructure/programs/shm/firewall.py                            34     25    26%
data_safe_haven/infrastructure/programs/shm/monitoring.py                          45     35    22%
data_safe_haven/infrastructure/programs/shm/networking.py                          41     32    22%
data_safe_haven/infrastructure/programs/sre/__init__.py                             0      0   100%
data_safe_haven/infrastructure/programs/sre/application_gateway.py                 24     14    42%
data_safe_haven/infrastructure/programs/sre/apt_proxy_server.py                    29     19    34%
data_safe_haven/infrastructure/programs/sre/backup.py                              18     11    39%
data_safe_haven/infrastructure/programs/sre/data.py                                84     69    18%
data_safe_haven/infrastructure/programs/sre/database_servers.py                    26     18    31%
data_safe_haven/infrastructure/programs/sre/dns_server.py                          38     26    32%
data_safe_haven/infrastructure/programs/sre/gitea_server.py                        46     35    24%
data_safe_haven/infrastructure/programs/sre/hedgedoc_server.py                     42     30    29%
data_safe_haven/infrastructure/programs/sre/identity.py                            32     23    28%
data_safe_haven/infrastructure/programs/sre/monitoring.py                          18     10    44%
data_safe_haven/infrastructure/programs/sre/networking.py                         113    104     8%
data_safe_haven/infrastructure/programs/sre/remote_desktop.py                      48     36    25%
data_safe_haven/infrastructure/programs/sre/software_repositories.py               42     30    29%
data_safe_haven/infrastructure/programs/sre/user_services.py                       47     34    28%
data_safe_haven/infrastructure/programs/sre/workspaces.py                          68     50    26%
data_safe_haven/infrastructure/project_manager.py                                 220    132    40%
data_safe_haven/provisioning/__init__.py                                            2      0   100%
data_safe_haven/provisioning/sre_provisioning_manager.py                           48     34    29%
data_safe_haven/resources/__init__.py                                               3      0   100%
data_safe_haven/serialisers/__init__.py                                             4      0   100%
data_safe_haven/serialisers/azure_serialisable_model.py                            22      0   100%
data_safe_haven/serialisers/context_base.py                                        14      2    86%
data_safe_haven/serialisers/yaml_serialisable_model.py                             40      0   100%
data_safe_haven/types/__init__.py                                                   4      0   100%
data_safe_haven/types/annotated_types.py                                           16      0   100%
data_safe_haven/types/enums.py                                                     10      0   100%
data_safe_haven/types/types.py                                                      2      0   100%
data_safe_haven/utility/__init__.py                                                 5      0   100%
data_safe_haven/utility/directories.py                                              8      0   100%
data_safe_haven/utility/file_reader.py                                             20      9    55%
data_safe_haven/utility/logger.py                                                 108     78    28%
data_safe_haven/utility/singleton.py                                                8      0   100%
data_safe_haven/validators/__init__.py                                              3      0   100%
data_safe_haven/validators/typer.py                                                19      0   100%
data_safe_haven/validators/validators.py                                           49      0   100%
data_safe_haven/version.py                                                          2      0   100%
---------------------------------------------------------------------------------------------------
TOTAL                                                                            4551   2760    39%
```

### :microscope: Tests

Tested locally

### CODE REVIEW

**Comment by jemrobinson** [comment]
I actually think this might be better as part of the main test function.

@JimMadge: Is there any reason we **wouldn't** want coverage to run normally?
@craddm: Can you add some example output to the summary of this PR?

**Comment by JimMadge** [comment]
I would keep it as two scripts.

You might not always want a coverage report when running tests. I would also guess there is some overhead to analysing which lines are run too.

If we add coverage we should run it in CI and (ideally) have reports in PRs. I think Codecov is no longer free for open source projects. Are there alternatives? GH Actions?

**Comment by jemrobinson** [comment]
[coverage.py](https://coverage.readthedocs.io/en/7.5.1/) is still free I think? I don't mind so much about whether it runs on every test run but I **do** want to get one of those nice bot comments on PRs that show you how you're doing with increased/decreased coverage. Anyone got a preferred way of doing that?

**Comment by craddm** [comment]
Codecov *is* still free for open-source, according to their [pricing page](https://about.codecov.io/pricing/), so could look at getting that going. That'd be able to do exactly what you're asking for James. Otherwise I can see if there's a simple way to do it using `coverage.py` or what have you

EDIT: although looking more into it, the caveat seems to be that it is free for "Open Source and Single Developer projects", and it's not clear whether *both* of those things need to be true. Ah, and also not clear if it does the patch coverage analysis for free any more... been a while since I've used it. Will dig into it more

**Comment by JimMadge** [comment]
I liked Codecov when I used it previously. Nice visualisations of how coverage breaks down per file, directory.

@craddm we might want to make the default output a little more concise for when we are working in the terminal.

**Comment by craddm** [comment]
> I liked Codecov when I used it previously. Nice visualisations of how coverage breaks down per file, directory.
> 
> @craddm we might want to make the default output a little more concise for when we are working in the terminal.

Also previously liked Codecov a lot. I've changed this branch a little to use `coverage.py` and a Github action that adds a comment to PRs to say how much code has changed/is covered, which is a lot more basic than what Codecov provides. See https://github.com/craddm/data-safe-haven/pull/18 for an example (where no code has changed, I'm just about to try modifying some just to see what happens)

I can still try to see if Codecov will work instead.

**Comment by JimMadge** [comment]
@craddm Happy to be guided by you. I think the essential thing is to have some kind of summary of coverage change in PRs.

**Comment by craddm** [comment]
Got a Codecov version working now too - example report should be viewable at https://app.codecov.io/gh/craddm/data-safe-haven/tree/use-codecov/data_safe_haven

It gives nice reports as comments on individual PRs too

**Comment by jemrobinson** [comment]
> EDIT: although looking more into it, the caveat seems to be that it is free for "Open Source and Single Developer projects", and it's not clear whether both of those things need to be true. Ah, and also not clear if it does the patch coverage analysis for free any more... been a while since I've used it. Will dig into it more

Looks like it's either/or (see screenshots below)

Screenshot 2024-05-08 at 16 33 06

Screenshot 2024-05-08 at 16 33 45

File: data_safe_haven/utility/logger.py [hunk 1]
```diff
@@ -59,6 +59,7 @@ def __init__(self, fmt: str, datefmt: str, *args: Any, **kwargs: Any):
             tracebacks_show_locals=True,
             **kwargs,
         )
+        print("hola chicos")
```
Comment (JimMadge) [inline data_safe_haven/utility/logger.py]: :eyes:

Comment (craddm) [reply data_safe_haven/utility/logger.py]: As you can see from the example comment here, https://github.com/craddm/data-safe-haven/pull/18#issuecomment-2100708446, the report from that bot does indeed provide links to where the coverage has changed and how!

Comment (JimMadge) [reply data_safe_haven/utility/logger.py]: Nice! I really like that comment. It is concise but has links to all the detail you would need :+1:.

Comment (JimMadge) [reply data_safe_haven/utility/logger.py]: That is just an action that works with `pytest-cov`?

That seems easier than making an account on Codecov.

Comment (craddm) [reply data_safe_haven/utility/logger.py]: Yes, I'm leaning towards this one for exactly that reason. Codecov was easy enough to do via my account, but then we are reliant on them and have some extra - albeit minor - management to do.

I have the relevant permission to add Codecov to our main repo, so I'm sure any of us can if we have an account. But it is an extra, external step

Comment (JimMadge) [reply data_safe_haven/utility/logger.py]: I would vote to go with the action then.

File: .github/workflows/test_code.yaml [hunk 2]
```diff
@@ -20,7 +28,22 @@ jobs:
...
-        run: hatch run test:test
+        run: hatch run test:cov
+
+      - name: Coverage comment
+        id: coverage_comment
+        uses: py-cov-action/python-coverage-comment-action@v3
+        with:
+          GITHUB_TOKEN: ${{ github.token }}
+
+      - name: Store Pull Request comment to be posted
```
Comment (jemrobinson) [inline .github/workflows/test_code.yaml]: Does this post the comment? If so maybe the name could be `Add coverage summary to pull request` or similar?

**Review by jemrobinson — APPROVED** [review]
LGTM. One minor comment

File: data_safe_haven/utility/logger.py [hunk 3]
```diff
@@ -59,7 +59,6 @@ def __init__(self, fmt: str, datefmt: str, *args: Any, **kwargs: Any):
             tracebacks_show_locals=True,
             **kwargs,
         )
-        print("hola chicos")
```
Comment (jemrobinson) [inline data_safe_haven/utility/logger.py]: ;(

File: pyproject.toml [hunk 4]
```diff
@@ -109,11 +109,22 @@ all = [
 dependencies = [
   "pytest~=8.1",
   "pytest-mock~=3.14",
+  "coverage~=7.5.1"
```
Comment (jemrobinson) [inline pyproject.toml]:
```suggestion
  "coverage~=7.5.1",
  "pytest~=8.1",
  "pytest-mock~=3.14",
```

Comment (jemrobinson) [reply pyproject.toml]: Syntax highlighting is making this look very ugly, but it's just alphabetising the dependencies :)

**Comment by jemrobinson** [comment]
Is this ready to go @craddm ?

**Comment by craddm** [comment]
Should be good to go now yes, assuming @JimMadge is happy with the level of verbosity

**Pull request merged**